haystack/agents/base.py

```diff
@@ -60,6 +61,35 @@ def __init__(
         self.name = name
         self.pipeline_or_node = pipeline_or_node
         self.description = description
+        self.output_variable = output_variable
+
+    def run(self, tool_input: str, params: Optional[dict] = None) -> str:
+        # We can only pass params to pipelines but not to nodes
+        if isinstance(self.pipeline_or_node, (Pipeline, BaseStandardPipeline)):
+            result = self.pipeline_or_node.run(query=tool_input, params=params)
+        elif isinstance(self.pipeline_or_node, BaseRetriever):
+            result = self.pipeline_or_node.run(query=tool_input, root_node="Query")
+        else:
+            result = self.pipeline_or_node.run(query=tool_input)
+
+        # if result was returned by a node it is of type tuple. We use only the output but not the name of the output.
+        # if result was returned by a pipeline it is of type dict that we can use directly.
+        if isinstance(result, tuple):
+            result = result[0]
+        if isinstance(result, dict):
+            if result and self.output_variable not in result:
+                raise ValueError(
+                    f"Tool {self.name} returned result {result} but "
+                    f"output variable '{self.output_variable}' not found in."
+                )
+            result = result.get(self.output_variable)
+        if isinstance(result, list):
+            result = result[0] if result else []
+        if isinstance(result, Answer):
+            result = result.answer
+        if isinstance(result, Document):
+            result = result.content
```

Comment:
We can put these lines in a separated function `_process_result()` so we can easily test each branch of those `if`s.

Note that you're not using `else`, so you're going to evaluate all the conditions all the times - not a big deal but I'd fix it.